aiida/work/ports.py

```diff
@@ -31,10 +31,34 @@ def get_description(self):
 
         return description
 
+class WithSerializeFct(object):
+    """
+    A mixin that adds support for a serialization function which is automatically applied on inputs that are not AiiDA data types.
+    """
+    def __init__(self, *args, **kwargs):
+        serialize_fct = kwargs.pop('serialize_fct', None)
+        super(WithSerializeFct, self).__init__(*args, **kwargs)
+        self._serialize_fct = serialize_fct
 
-class InputPort(WithNonDb, ports.InputPort):
+    def serialize(self, value):
+        from aiida.orm import Data
+        if self._serialize_fct is None or isinstance(value, Data):
```

Comment:
I wouldn't have the exception of `Data`.  IMHO if there is a serialize function it should be used in any case i.e.
```
if self._serialize_fct is not None:
    return self._serialize_fct(value)
return value
```

Reply:
Again I would leave out `fct`

Reply:
I think the issue (if I remember the discussions correctly) with not making an exception for ``Data`` is that using the inputs as stored in the DB may not give the same result. Essentially, if I write 
```
res, node = run_get_node(Process, **inputs)
res2 = run(Process, **node.get_inputs_dict())
```
I would expect that ``res`` and ``res2`` are the same. Not allowing serialization on ``Data`` objects is an easy way to enforce this.

For the name of the input, I think it would be nice if there is some indication that a function (or callable) needs to be passed, not e.g. a boolean flag. Maybe ``serializer`` instead of ``serialize``?

Reply:
Re: no serialisation of data inputs. Sounds reasonable. But can we raise at that point to say that the user shouldn't have set a serialised on that port. Otherwise it just skips it silently which seems dangerous.

Serializer sounds good to me :)

Reply:
Doesn't that still leave the problem that the user won't be able to just re-create the process from the data as it is stored in the database? Maybe just a warning would be ok?

Re naming: done.